src/python/pants/backend/python/util_rules/local_dists.py

```diff
@@ -16,19 +15,79 @@
 from pants.backend.python.util_rules.pex_requirements import PexRequirements
 from pants.backend.python.util_rules.python_sources import PythonSourceFiles
 from pants.build_graph.address import Address
-from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.core.goals.package import BuiltPackage
+from pants.core.util_rules import archive
+from pants.core.util_rules.archive import UnzipBinary
 from pants.core.util_rules.source_files import SourceFiles
 from pants.engine.addresses import Addresses
-from pants.engine.fs import Digest, DigestContents, DigestSubset, MergeDigests, PathGlobs, Snapshot
+from pants.engine.fs import Digest, DigestSubset, MergeDigests, PathGlobs, Snapshot
+from pants.engine.process import BashBinary, Process, ProcessResult
 from pants.engine.rules import Get, MultiGet, collect_rules, rule
-from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest
+from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest, WrappedTarget
 from pants.util.dirutil import fast_relpath_optional
 from pants.util.docutil import doc_url
 from pants.util.meta import frozen_after_init
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True)
+class LocalDistWheels:
+    """Contains the wheels isolated from a single local Python distribution."""
+
+    wheel_paths: list[str]
```

Comment:
It may be true that wheel_paths are populates from a Snapshot below which happens to be sorted and it may be true that a set will have a stable order when formed multiple times in the same pantsd Python process - but both not encoding stable order and not encoding immutability in a "frozen" dataclass seems sub-optimal for sanity sake at the least.

Reply:
Yea... I'll do this for consistency's sake.

But given how tightly we've adhered to functional patterns throughout the system, I do sometimes wonder whether defensively freezing dataclasses is worth it. It's clearly necessary when something will be used in a `Get` (for `hash`), but not necessary when something is only the return value of a `@rule`.

Reply:
I'd argue if that's true its too subtle. If a rule returns a thing that thing ~must be used by another rule as an input and so you'd expect it better be immutable.

Put it another way, if I have to debug "strange" behavior in the engine, a 1st stop would be to look for violations of patterns in the Python rules. Being able to sanely rely on ~types would be a good 1st stop in that 1st stop. Seeing frozen != frozen would give me pause here without a comment.

Reply:
> Put it another way, if I have to debug "strange" behavior in the engine, a 1st stop would be to look for violations of patterns in the Python rules. Being able to sanely rely on ~types would be a good 1st stop in that 1st stop. Seeing frozen != frozen would give me pause here without a comment.

Sure. But we have had zero cases of those types of errors in the last few years... and whether that is due more to our defensive freezing of values, or due to mutating of a `@rule`-returned value just not being something that is likely/useful is the question that I'm referring to here.